### PR TITLE
Login logout

### PR DESCRIPTION
**Pull request by mariaefOT**
- Added login and logout components
- Changed role management
- Tests for these components have been added

### CODE REVIEW

File: src/CSS/Login.css [hunk 1]
```diff
@@ -0,0 +1,9 @@
+.login {
```
Comment (ghost) [inline src/CSS/Login.css]: I suggest you add the element to the name of the class. Example: `login-div`.

File: src/CSS/Login.css [hunk 2]
```diff
@@ -0,0 +1,9 @@
...
+    margin-top: 100px;
+}
+
+.conteiner {
```
Comment (ghost) [inline src/CSS/Login.css]: container*

File: src/api/index.js [hunk 3]
```diff
@@ -1,34 +1,57 @@
...
+    }
+  );
+
+  return response.data.authToken;
+};
+
+export const getRole = (token) => {
   return(
-    axiosClient.get(`${role === 'USER' ? 'user/' : ''}account/checking`, {
+    axiosClient.get('user/role', {
```
Comment (ghost) [inline src/api/index.js:20]: Did you try replacing `${token}` with `${sessionStorage.getItem("token")}` here?

File: src/pages/LoginScreen.js [hunk 4]
```diff
@@ -0,0 +1,14 @@
+import Login from "../components/Login";
+import '../CSS/Login.css';
+
+const LoginScreen = () => {
```
Comment (ghost) [inline src/pages/LoginScreen.js]: If you follow the naming conventions for pages, you should name it `LoginPage` instead.

File: src/validations/validateLogin.js [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+const validateUserName = (username) => {
+    const regex =  /^[^@]+@[^@]+\.[a-zA-Z]{2,}$/;
+    return !!username && regex.test(username);
+}
+
+const validatePassword = (pass) => {
+    return !!pass && pass.length >= 8;
+}
+
+export const validateLoginForm = (values) => {
```
Comment (ghost) [inline src/validations/validateLogin.js]: "credentials" sounds like a better and more specific name for the parameter of this function.

File: src/components/Login.js [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+import { Form, Button, Alert, Badge } from 'react-bootstrap';
+import { authenticateUser, getRole } from "../api";
+import { useNavigate } from 'react-router-dom';
+import { validateLoginForm } from "../validations/validateLogin";
+
+const Login = () => {
+    let navigate = useNavigate(); 
+    const [validated, setValidated] = useState(false);
+    const [show, setShow] = useState(false);
+    const [values, setValues] = useState({
```
Comment (ghost) [inline src/components/Login.js]: "Values" could refer to any values (even ones that have nothing to do with the login functionality). I suggest naming it _credentials_.

File: src/components/Login.js [hunk 7]
```diff
@@ -0,0 +1,108 @@
+import { useState } from "react";
+import { Form, Button, Alert, Badge } from 'react-bootstrap';
+import { authenticateUser, getRole } from "../api";
+import { useNavigate } from 'react-router-dom';
+import { validateLoginForm } from "../validations/validateLogin";
+
+const Login = () => {
+    let navigate = useNavigate(); 
+    const [validated, setValidated] = useState(false);
+    const [show, setShow] = useState(false);
```
Comment (ghost) [inline src/components/Login.js]: Choose a more specific name. For example: `showErrorMessage`. If you name it "show", a person who reads the code for the first time won't know what you meant or what you truly wanted to do with that state.

File: src/components/Login.js [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+                getRole(token).then((response) => {
+                    const role = `${response.data[0].authority} ${response.data[1].authority}`;
+                    if(role.includes('ADMIN')){
+                        sessionStorage.setItem('role','ADMIN');
+                    }
+                    if(role.includes('USER')){
+                        sessionStorage.setItem('role','USER'); 
+                    }
+                    navigate('/home'); 
+                });
```
Comment (ghost) [inline src/components/Login.js]: You can split these steps into:

 1) a `login` function, which you can put in a `login.js` file inside the `api` folder.
 2) a function that sets a given role in the `sessionStorage`.

... so you can reduce the amount of logic the `Login` component has + reduce the amount of nested `if` statements which make your code less maintainable.

File: src/tests/Login.test.js [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+                <Login/>
+            </Router>
+        );
+        expect(view).toBeDefined();
+
+        expect(screen.getByLabelText(/User Name/i)).toBeTruthy();
+        expect(screen.getByLabelText(/Password/i)).toBeTruthy();
+    });
+
+    it('Should onSubmit is called when all fields pass validation and redirect', () => {
```
Comment (ghost) [inline src/tests/Login.test.js]: Should log in when the "Sign In" button is pressed*

File: src/tests/Login.test.js [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+        const userName = screen.getByRole('textbox', {name: /user name/i });
```
Comment (ghost) [inline src/tests/Login.test.js]: Remember that `screen.getByRole()` returns an element, so the name of the `const` should tell that. If you name it _username_, you imply that the return value is a string when it's not.

You should call it: _usernameTxtBox_, _usernameInput_, or something like that. ;)

File: src/tests/Login.test.js [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+
+        user.click(screen.getByRole('button', {name: /SIGN IN/i }));
+
+        waitFor(() => {
+            expect(onSubmit).toHaveBeenCalledTimes(1);
+            expect(mockedUsedNavigate).toHaveBeenCalledWith("/home");
+        });
+    });
+
+    it('Required fields must be displayed and not redirect', () => {
```
Comment (ghost) [inline src/tests/Login.test.js]: Two options here:

1) "Should show an error message when the credentials are invalid".
2) "Should show the 'Failed to login!' error message when the credentials are invalid".

Comment (ghost) [reply src/tests/Login.test.js]: In this case, you can also say "...when the login form fields are empty"

File: src/tests/Logout.test.js [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+    const yesModalBtn = jest.fn();
+    const noModalBtn = jest.fn();
+    const mockedUsedNavigate = jest.fn();
+
+    jest.mock("react-router-dom", () => ({
+        ...jest.requireActual("react-router-dom"),
+        useNavigate: () => mockedUsedNavigate,
+    }));
+
+    it('Should logout with successfully and redirect to the login after', async () => {
```
Comment (ghost) [inline src/tests/Logout.test.js]: Should log out successfully after pressing the logout confirmation button*

**Review by ghost — CHANGES_REQUESTED** [review]

File: src/CSS/Login.css (hunk 1 above)
Comment (mariaefOT) [reply src/CSS/Login.css]: Done

File: src/CSS/Login.css (hunk 2 above)
Comment (mariaefOT) [reply src/CSS/Login.css]: Done

File: src/api/index.js (hunk 3 above)
Comment (mariaefOT) [reply src/api/index.js:20]: I missed it, thanks

File: src/pages/LoginScreen.js (hunk 4 above)
Comment (mariaefOT) [reply src/pages/LoginScreen.js]: Done

File: src/validations/validateLogin.js (hunk 5 above)
Comment (mariaefOT) [reply src/validations/validateLogin.js]: Done

File: src/components/Login.js (hunk 6 above)
Comment (mariaefOT) [reply src/components/Login.js]: Thanks for the suggestion, I'll take it.

File: src/components/Login.js (hunk 7 above)
Comment (mariaefOT) [reply src/components/Login.js]: Done

File: src/tests/Login.test.js (hunk 9 above)
Comment (mariaefOT) [reply src/tests/Login.test.js]: Done

File: src/tests/Logout.test.js (hunk 12 above)
Comment (mariaefOT) [reply src/tests/Logout.test.js]: Done

File: src/tests/Login.test.js (hunk 11 above)
Comment (mariaefOT) [reply src/tests/Login.test.js]: Done

File: src/tests/Login.test.js (hunk 10 above)
Comment (mariaefOT) [reply src/tests/Login.test.js]: Done

File: src/components/Login.js (hunk 8 above)
Comment (mariaefOT) [reply src/components/Login.js]: Try to apply. Any mistake tell me

File: src/tests/Login.test.js (hunk 10 above)
Comment (ghost) [reply src/tests/Login.test.js]: The same comment applies to the password.

**Review by ghost — APPROVED** [review]

**Pull request merged**